File: database/foxx/backend/src/model/documents/CommentBucket.ts
import { Attribute, Document, Entity, Index, Nested } from 'type-arango';

export enum CommentSource {
    GOOGLE_PLACES = 'GOOGLE_PLACES',
    TRIPADVISOR = 'TRIPADVISOR',
}

@Nested()
export class Comment {
    @Attribute()
    rating: number;

    @Attribute()
    text: string;

    @Attribute()
    timestamp: Date;
}

@Document()
export class CommentBucket extends Entity {
    @Index('persistent')
    @Attribute()
    cenoteId: string;

    @Attribute()
    source: CommentSource;

    @Attribute()
    url: string;

    @Attribute()
    count: number;

    @Attribute()
    sumRating: number;

    @Attribute()
    comments: Comment[];
}

File: backend/src/main/java/org/cenoteando/models/CommentBucket.java
package org.cenoteando.models;

import com.arangodb.springframework.annotation.ArangoId;
import com.arangodb.springframework.annotation.Document;
import java.util.Date;
import java.util.List;

import lombok.Getter;
import lombok.Setter;
import org.springframework.data.annotation.Id;
import org.springframework.data.annotation.Persistent;

@Document("Comments")
@Getter
@Setter
public class CommentBucket {

    public enum CommentSource {
        GOOGLE_PLACES,
        TRIPADVISOR,
    }

    @Id // db document field: _key
    private String id;

    @ArangoId // db document field: _id
    private String arangoId;

    @Persistent
    private String cenoteId;

    private CommentSource source;
    private int count;
    private int sumRating;
    private List<Comment> comments;

    public CommentBucket(
        String cenoteId,
        CommentSource source,
        int count,
        List<Comment> comments
    ) {
        this.cenoteId = cenoteId;
        this.source = source;
        this.count = comments.size();
        this.comments = comments;
        this.sumRating = 0;
        for (Comment comment : comments) this.sumRating += comment.getRating();
    }

}
